OrbitCore/LinuxPerfEventProcessor.h

```diff
@@ -0,0 +1,83 @@
+//-----------------------------------
+// Copyright Pierric Gimmig 2013-2019
+//-----------------------------------
+#pragma once
+
+#if __linux__
+
+#include <queue>
+#include "LinuxPerfEvent.h"
+#include "LinuxPerfEventVisitor.h"
+
+// A comparator used for the priority queue, such that pop/top will
+// always return the oldest event in the queue.
+class TimeStampCompare
+{
+public:
+    bool operator() (std::shared_ptr<LinuxPerfEvent> lhs, std::shared_ptr<LinuxPerfEvent> rhs)
+    {
+        return lhs->Timestamp() > rhs->Timestamp();
+    }
+};
+
+class LinuxPerfEventProcessor
+{
+public:
+    // While processing, we do not touch the events with a timestamp less 
+    // then 1/100 sec smaller than the most recent one in the queue.
+    // This way we can ensure, that all events (from different sources)
+    // are processed in the correct order.
+    uint64_t OFFSET = 10000000 /*ns*/;
+
+    LinuxPerfEventProcessor(std::unique_ptr<LinuxPerfEventVisitor> a_Visitor) :
+        m_Visitor(std::move(a_Visitor))
+    { 
+    }
+
+    void Push(std::shared_ptr<LinuxPerfEvent> e)
+    {
+        uint64_t timestamp = e->Timestamp();
+        if (timestamp > m_MaxTimestamp)
+            m_MaxTimestamp = timestamp;
+        m_EventQueue.push(e);
+    }
+
+    void ProcessAll()
+    {
+        while(!m_EventQueue.empty())
+        {
+            std::shared_ptr<LinuxPerfEvent> event = m_EventQueue.top();
+            event->accept(m_Visitor.get());
+            m_EventQueue.pop();
+        }
+    }
+
+    void ProcessTillOffset()
+    {
+        // Process the events in the event_queue
+        while(!m_EventQueue.empty())
+        {
+            std::shared_ptr<LinuxPerfEvent> event = m_EventQueue.top();
+
+            // We should not read all events, otherwise we could miss events
+            // close to the max timestamp in the queue.
+            if (event->Timestamp() + OFFSET > m_MaxTimestamp) 
+                break;
+            event->accept(m_Visitor.get());
```

Comment:
event->accept(visitor) ends up calling visitor->visit(event).  Could we call visitor->visit(event) directly instead of 2 virtual non inlined function calls for every event?  I'm not convinced of the added value of using a visitor pattern here.  Are you planning on having different visitor types?

Reply:
Due to the polymorphism, we can not directly call `visitor->visit(event)`. At the time of calling the visit/accept method, we do not know about the type of the event anymore. That's the point of the visitor patter. I choosed it, as we can indeed implement different visitors, e.g. one for sampling, one for uprobes and one for sched_events.
I mean, we could also add a virtual method "process" on the event's base class and then on each concrete class, do the interaction with the OrbitApp directly there. But then, it might be hard to keep track of state (e.g. for the callstacks of uprobes that belong to a certain function). 
I like the pattern here, as we would have a clean cut between the abstract representation of different records for perf events and processing them. Also from what I experienced so far, the two virtual calls would not hurt us that much. 

An alternative to the visitor pattern with non-virtual (or only one virtual call), would be to add an EVENT_TYPE enum value to each perf_event class and do a switch on the different Types.
But as I said, in my opinion this pattern is nice, to keep the code clean and reusable here.

Reply:
See https://github.com/pierricgimmig/orbitprofiler/pull/82#discussion_r348802270.